types/wlr_output.c

```diff
@@ -666,6 +673,32 @@ struct wlr_output *wlr_output_from_resource(struct wl_resource *resource) {
 	return wl_resource_get_user_data(resource);
 }
 
+static void output_cursor_damage_whole(struct wlr_output_cursor *cursor);
+
+void wlr_output_lock_software_cursors(struct wlr_output *output, bool lock) {
+	if (lock) {
+		++output->software_cursor_locks;
+	} else {
+		assert(output->software_cursor_locks > 0);
+		--output->software_cursor_locks;
+	}
+	wlr_log(WLR_DEBUG, "%s hardware cursors on output '%s' (locks: %d)",
+		lock ? "Disabling" : "Enabling", output->name,
+		output->software_cursor_locks);
+
+	if (output->software_cursor_locks > 0 && output->hardware_cursor != NULL) {
+		assert(output->impl->set_cursor);
+		output->impl->set_cursor(output, NULL, 1,
+			WL_OUTPUT_TRANSFORM_NORMAL, 0, 0, true);
+		output_cursor_damage_whole(output->hardware_cursor);
+		output->hardware_cursor = NULL;
+	}
+
+	// If it's possible to use hardware cursors again, don't switch immediately
+	// since a recorder is likely to lock software cursors for the next frame
+	// again.
```

Comment:
Well, when are hardware cursors restored then?

Reply:
The next time the cursor is updated again.

Reply:
Doesn't sound very robust. Could the cursor perhaps be updated after a client detaches from the manager interface?

Reply:
This would complicate this already-complicated code. It's not like using software cursors for a few seconds is the end of the world.

Reply:
It could be an entire eternity if for instance this is a fullscreen shell with a single output. I don't think it would add more than a few lines.

Reply:
Doing this would require to track bound globals (ie. allocate a per-bound-global struct to contain the lock state), which we don't currently do. I prefer to keep this as a follow-up issue since we don't support fullscreen shell (so the situation you describe is very unlikely to happen).

Reply:
The situation is very likely to happen even in sway, all you have to do is not move your mouse over another window after you stop capturing. Not restoring the state and requiring compositors to hack around this for you is a hack in of itself. Moreover, the function doesn't actually do what it states it should do, and changing this behavior in the future could be an API break.

Reply:
Even if you don't move your mouse over another client, you're very likely to update it anyway by moving over a text input or a link. We're of course not requiring compositors to hack around this. This API is nowhere near stable so I'm not concerned about changing its semantics.

Reply:
>It's not like using software cursors for a few seconds is the end of the world.

This is why I don't care

Reply:
It would have been absolutely no problem if it was just a few seconds, as it stands though it could go to infinity in case of e.g. embedded, single program systems where the cursor doesn't change if it goes over text or not.

Reply:
Do we really care about recording the screen with the cursor on such systems?

Also support for writeback DRM feature will allow us to stop doing that on the DRM backend (and drivers that support it).